oak_crypto/src/hpke.rs

```diff
@@ -0,0 +1,259 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{
+    aead::{AeadKey, AeadNonce, AEAD_ALGORITHM_KEY_SIZE_BYTES, AEAD_NONCE_SIZE_BYTES},
+    schema::AeadEncryptedMessage,
+    util::{i2osp, xor},
+};
+use alloc::vec::Vec;
+use anyhow::{anyhow, Context};
+use hpke::{
+    aead::{AeadCtxR, AeadCtxS, AesGcm256},
+    kdf::HkdfSha256,
+    kem::DhP256HkdfSha256,
+    Deserializable, Kem as KemTrait, OpModeR, OpModeS, Serializable,
+};
+use rand_core::OsRng;
+
+type Aead = AesGcm256;
+type Kdf = HkdfSha256;
+type Kem = DhP256HkdfSha256;
+pub(crate) type PrivateKey = <Kem as KemTrait>::PrivateKey;
+pub(crate) type PublicKey = <Kem as KemTrait>::PublicKey;
+pub(crate) type EncappedKey = <Kem as KemTrait>::EncappedKey;
+
+/// Maximum sequence number which can fit in [`AEAD_NONCE_SIZE_BYTES`] bytes.
+/// https://www.rfc-editor.org/rfc/rfc9180.html#name-encryption-and-decryption
+const MAX_SEQUENCE_NUMBER: u128 = (1 << (8 * AEAD_NONCE_SIZE_BYTES)) - 1;
+
+pub(crate) struct KeyPair {
+    pub(crate) private_key: PrivateKey,
+    pub(crate) public_key: PublicKey,
+}
+
+impl KeyPair {
+    /// Randomly generates a key pair.
+    pub(crate) fn generate() -> Self {
+        let (private_key, public_key) = Kem::gen_keypair(&mut OsRng);
+        Self {
+            private_key,
+            public_key,
+        }
+    }
+
+    /// Returns a NIST P-256 SEC1 encoded point public key.
+    /// <https://secg.org/sec1-v2.pdf>
+    pub fn get_serialized_public_key(&self) -> Vec<u8> {
+        self.public_key.to_bytes().to_vec()
+    }
+}
+/// Sets up an HPKE sender by generating an ephemeral keypair (and serializing the corresponding
+/// public key) and creating a sender context.
+/// <https://www.rfc-editor.org/rfc/rfc9180.html#name-encryption-to-a-public-key>
+pub(crate) fn setup_base_sender(
+    serialized_recipient_public_key: &[u8],
+    info: &[u8],
+) -> anyhow::Result<(Vec<u8>, SenderContext, SenderResponseContext)> {
+    let recipient_public_key = PublicKey::from_bytes(serialized_recipient_public_key)
+        .map_err(|error| anyhow!("couldn't deserialize recipient public key: {}", error))?;
+
+    let (encapped_key, sender_context) = hpke::setup_sender::<Aead, Kdf, Kem, _>(
+        &OpModeS::Base,
+        &recipient_public_key,
+        info,
+        &mut OsRng,
+    )
+    .map_err(|error| anyhow!("couldn't create sender context: {}", error))?;
+
+    let mut response_key = [0u8; AEAD_ALGORITHM_KEY_SIZE_BYTES];
```

Comment:
It feels like this could be more Rust-y. I'm not sure what's the best approach, though.

Option 1: use `Default::default()`, but that would give the impression that a zero key is a reasonable default.
Option 2: have something like `AeadKey::zero()`, which means you don't have to drag the constant around everywhere
Option 3: I don't know the APIs enough, but instead of first creating an emoty key and then calling `export`, see if `export` itself could return you the tkey.